File: linked-list-shift/LinkedListShift.java
import java.util.*;

class LinkedListShift {
  public static Node shiftLinkedList(Node head, int k) {
    Node tail = head;
    int length = 1;
    while(tail.next != null) {
        length += 1;
        tail = tail.next;
    }

    int offset = Math.abs(k) % length;
    if(offset == 0) return head;
    
    int position = k > 0 ? length - offset : offset;
    Node newTail = head;
    
    for(int i = 1; i < position; i++) {
        newTail = newTail.next;
    }

    Node newHead = newTail.next;
    newTail.next = null;
    tail.next = head;
    return newHead;
  }

  static class Node {
    public int value;
    public Node next;

    public Node(int value) {
      this.value = value;
      next = null;
    }
  }
}

File: linked-list-shift/LinkedListShift.swift
class LinkedListShift {
    class Node {
        var value: Int
        var next: Node?
        init(value: Int){
            self.value = value
        }
    }

    func linkedListShift(_ head: Node, _ k: Int) -> Node {
        var tail: Node? = head
        var length = 1
        while tail!.next != nil {
            tail = tail!.next
            length += 1
        }

        var offset = abs(k) % length
        if(offset == 0) { return head }
        var position = (k > 0) ? length - offset : offset;

        var newTail: Node? = head
        for _ in stride(from: 1, to: position, by: 1) {
            newTail = newTail!.next
        }

        let newHead = newTail!.next
        newTail!.next = nil
        tail!.next = head
        
        return newHead
    }
}